client/benchmark/src/main/java/org/elasticsearch/client/benchmark/AbstractBenchmark.java

```diff
@@ -0,0 +1,88 @@
+import org.elasticsearch.client.benchmark.ops.bulk.BulkRequestExecutor;
+import org.elasticsearch.client.benchmark.ops.search.SearchBenchmarkTask;
+import org.elasticsearch.client.benchmark.ops.search.SearchRequestExecutor;
+import org.elasticsearch.common.SuppressForbidden;
+
+import java.io.Closeable;
+
+public abstract class AbstractBenchmark<T extends Closeable> {
+    private static final int SEARCH_BENCHMARK_ITERATIONS = 10_000;
+
+    protected abstract T client(String benchmarkTargetHost) throws Exception;
+
+    protected abstract BulkRequestExecutor bulkRequestExecutor(T client, String indexName, String typeName);
+
+    protected abstract SearchRequestExecutor searchRequestExecutor(T client, String indexName);
+
+    @SuppressForbidden(reason = "system out is ok for a command line tool")
+    public final void run(String[] args) throws Exception {
+        if (args.length < 6) {
+            System.err.println(
+                "usage: benchmarkTargetHostIp indexFilePath indexName typeName numberOfDocuments bulkSize [search request body]");
+            System.exit(1);
+        }
+        String benchmarkTargetHost = args[0];
+        String indexFilePath = args[1];
+        String indexName = args[2];
+        String typeName = args[3];
+        int totalDocs = Integer.valueOf(args[4]);
+        int bulkSize = Integer.valueOf(args[5]);
+
+        int totalIterationCount = (int) Math.floor(totalDocs / bulkSize);
+        // consider 40% of all iterations as warmup iterations
+        int warmupIterations = (int) (0.4d * totalIterationCount);
+        int iterations = totalIterationCount - warmupIterations;
+        String searchBody = (args.length == 7) ? args[6] : null;
+
+        T client = client(benchmarkTargetHost);
+
+        BenchmarkRunner benchmark = new BenchmarkRunner(warmupIterations, iterations,
+            new BulkBenchmarkTask(
+                bulkRequestExecutor(client, indexName, typeName), indexFilePath, warmupIterations + iterations, bulkSize));
+
+        try {
+            benchmark.run();
+            if (searchBody != null) {
+                for (int run = 1; run <= 5; run++) {
+                    System.out.println("=============");
+                    System.out.println(" Trial run " + run);
+                    System.out.println("=============");
+
+                    for (int throughput = 100; throughput <= 100_000; throughput *= 10) {
+                        //force a GC between trials
+                        System.gc();
+
+                        BenchmarkRunner searchBenchmark = new BenchmarkRunner(SEARCH_BENCHMARK_ITERATIONS, SEARCH_BENCHMARK_ITERATIONS,
+                            new SearchBenchmarkTask(
+                                searchRequestExecutor(client, indexName), searchBody, 2 * SEARCH_BENCHMARK_ITERATIONS, throughput));
+                        System.out.printf("Target throughput = %d ops / s%n", throughput);
```

Comment:
I could give you a *veeery* long explanation on that one but I try to keep it short. :)

Actually, you're asking two questions:

**1.) What does "target throughput" mean?**

It means that we are aiming to achieve a certain rate of operations / second. An operation in this context is usually one request we issue with the REST client. E.g., if you issue one search request, that's one operation. If you issue one bulk request (with N documents), that's also one operation.

Now, we could just run in a tight loop and issue as many requests as we can possibly do but this has an adverse effect on latency (and also normally one shouldn't run systems at 100% utilization for that reason).

**2.) Why does the actual throughput not match the target throughput?**

This has multiple reasons:

* The client is saturated and cannot keep up with the target throughput. Imagine we want to run at 1000 operations per second but each operation takes 5ms. You can only do 1s / 5ms = 200 operations per second so this target throughput is not achievable.

* Measurement approach: We could inject operations at an arbitrary target throughput into the system (by creating them on a defined schedule in a dedicated thread). But then they would just pile up in a queue before the client is able to handle them and would have a negative effect on latency (but not at all on service time). Tbh. this is a vast topic and it is hard to explain in just a couple of sentences. If you can spare an hour and are interested, you can watch the talk [How not to measure latency](https://www.youtube.com/watch?v=lJ8ydIuPFeU) by Gil Tene on Youtube.

* To a lesser degree this also benchmark driver overhead (and this should get even a bit better with JLBH) or rather due to scheduling jitter.

Does this make more sense now?